conjure_oxide/tests/generated_tests.rs

```diff
@@ -299,44 +299,47 @@ fn integration_test_inner(
 }
 
 fn assert_vector_operators_have_partially_evaluated(model: &conjure_core::Model) {
-    model.constraints.transform(Arc::new(|x| {
-        use conjure_core::ast::Expression::*;
-        match &x {
-            Bubble(_, _, _) => (),
-            Atomic(_, _) => (),
-            Sum(_, vec) => assert_constants_leq_one(&x, vec),
-            Min(_, vec) => assert_constants_leq_one(&x, vec),
-            Max(_, vec) => assert_constants_leq_one(&x, vec),
-            Not(_, _) => (),
-            Or(_, vec) => assert_constants_leq_one(&x, vec),
-            And(_, vec) => assert_constants_leq_one(&x, vec),
-            Eq(_, _, _) => (),
-            Neq(_, _, _) => (),
-            Geq(_, _, _) => (),
-            Leq(_, _, _) => (),
-            Gt(_, _, _) => (),
-            Lt(_, _, _) => (),
-            SafeDiv(_, _, _) => (),
-            UnsafeDiv(_, _, _) => (),
-            SumEq(_, vec, _) => assert_constants_leq_one(&x, vec),
-            SumGeq(_, vec, _) => assert_constants_leq_one(&x, vec),
-            SumLeq(_, vec, _) => assert_constants_leq_one(&x, vec),
-            DivEqUndefZero(_, _, _, _) => (),
-            Ineq(_, _, _, _) => (),
-            // this is a vector operation, but we don't want to fold values into each-other in this
-            // one
-            AllDiff(_, _) => (),
-            WatchedLiteral(_, _, _) => (),
-            Reify(_, _, _) => (),
-            AuxDeclaration(_, _, _) => (),
-            UnsafeMod(_, _, _) => (),
-            SafeMod(_, _, _) => (),
-            ModuloEqUndefZero(_, _, _, _) => (),
-            Neg(_, _) => (),
-            Minus(_, _, _) => (),
-        };
-        x.clone()
-    }));
+    let expressions = &model.constraints;
+    for expression in expressions {
+        // Collect all nodes using `universe` and process bottom-up
+        for node in expression.universe().iter().rev() {
```

Comment:
oh intriguing! why in reverse?

try `for expression in universe_bi(model.constraints)` and see if that works too btw. so you would remove the nested for loops and have only one.

see #544 for and the example I linked to in that issue for the exact syntax.

Reply:
In reverse because I thought that .universe would retrieve elements from node not in depth-first manner. However, I double checked the implementation, and it looks that it does do it in depth-first order

Reply:
Actually, I've taken a second look at it this morning, and I figured it out. @ozgurakgun Please review :)